extensions/amp-app-banner/0.1/amp-app-banner.js

```diff
@@ -0,0 +1,405 @@
+import {viewportFor} from '../../../src/viewport';
+import {vsyncFor} from '../../../src/vsync';
+import {CSS} from '../../../build/amp-app-banner-0.1.css';
+import {documentInfoFor} from '../../../src/document-info';
+import {xhrFor} from '../../../src/xhr';
+import {assertHttpsUrl} from '../../../src/url';
+import {isExperimentOn} from '../../../src/experiments';
+import {removeElement, openWindowDialog} from '../../../src/dom';
+import {storageFor} from '../../../src/storage';
+import {timer} from '../../../src/timer';
+import {parseUrl} from '../../../src/url';
+
+
+const TAG = 'amp-app-banner';
+
+
+/**
+ * @private visible for testing.
+ */
+export class AbstractAppBanner extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @protected */
+  setupOpenLink_(openInAppUrl, installAppUrl) {
+    const openLink = this.element.querySelector('a[open-link]');
+    openLink.addEventListener('click', () => {
+      this.openLinkClicked_(openInAppUrl, installAppUrl);
+    });
+  }
+
+
+  /** @private */
+  openLinkClicked_(openInAppUrl, installAppUrl) {
+    timer.delay(() => {
+      this.redirectTopLocation_(installAppUrl);
+    }, 1500);
+    openWindowDialog(this.win, openInAppUrl, '_top');
+  }
+
+
+  /** @private */
+  redirectTopLocation_(link) {
+    this.win.top.location.replace(link);
+  }
+
+  /**
+   * Creates and append a close button.
+   * @protected
+   */
+  addDismissButton_() {
+    const dismissButton = this.win.document.createElement('button');
+    dismissButton.classList.add('amp-app-banner-dismiss-button');
+    dismissButton.setAttribute('aria-label',
+        this.element.getAttribute('data-dismiss-button-aria-label') ||
+        'Dismiss');
+    const boundOnDismissButtonClick = this.onDismissButtonClick_.bind(this);
+    dismissButton.addEventListener('click', boundOnDismissButtonClick);
+    this.element.appendChild(dismissButton);
+  }
+
+  /**
+   * Dismisses the app banner and persist it.
+   * @protected
+   */
+  onDismissButtonClick_() {
+    vsyncFor(this.win).run({
+      measure: null,
+      mutate: handleDismiss,
+    }, {
+      element: this.element,
+      viewport: viewportFor(this.win),
+      storagePromise: storageFor(this.win),
+      storageKey: this.getStorageKey_(),
+    });
+  }
+
+  /** @private */
+  getStorageKey_() {
+    const elementId = user.assert(this.element.id,
+        'amp-app-banner should have an id.');
+    return 'amp-app-banner:' + elementId;
+  }
+
+  /** @protected */
+  isDismissed() {
+    return storageFor(this.win)
+        .then(storage => storage.get(this.getStorageKey_()))
+        .then(persistedValue => !!persistedValue, reason => {
+          dev.error(TAG, 'Failed to read storage', reason);
+          return false;
+        });
+  }
+
+  /** @protected */
+  checkIfDismissed_() {
+    this.isDismissed().then(isDismissed => {
+      if (isDismissed) {
+        this.hide_();
```

Comment:
That's not the case actually. But I think it should be. Will update to reflect this.